structured/privilege.go

```diff
@@ -0,0 +1,258 @@
+		return p.Users[i].User >= user
+	})
+	if idx == len(p.Users) || p.Users[idx].User != user {
+		// Not found.
+		return
+	}
+
+	copy(p.Users[idx:], p.Users[idx+1:])
+	p.Users[len(p.Users)-1] = nil
+	p.Users = p.Users[:len(p.Users)-1]
+}
+
+// DefaultDatabasePrivilegeDescriptor returns a privilege descriptor
+// with ALL privileges for the root user.
+func DefaultDatabasePrivilegeDescriptor() *PrivilegeDescriptor {
+	return &PrivilegeDescriptor{
+		Users: []*UserPrivileges{
+			{
+				User:       security.RootUser,
+				Privileges: (1 << uint32(Privilege_ALL)),
+			},
+		},
+	}
+}
+
+// Grant adds new privileges to this descriptor for a given list of users.
+// TODO(marc): if all privileges other than ALL are set, should we collapse
+// them into ALL?
+func (p *PrivilegeDescriptor) Grant(user string, privileges PrivilegeList) {
+	userPriv := p.findOrCreateUser(user)
+	for _, p := range privileges {
+		if p == Privilege_ALL {
+			// TODO(marc): even though this should not happen, error out if
+			// ALL is not the only privilege being granted.
+			userPriv.Privileges = (1 << uint32(p))
+			break
+		} else if userPriv.Privileges&(1<<uint32(Privilege_ALL)) != 0 {
+			// If we have ALL, don't set others.
+			continue
+		}
+		userPriv.Privileges |= (1 << uint32(p))
+	}
+}
+
+// Revoke removes privileges from this descriptor for a given list of users.
+func (p *PrivilegeDescriptor) Revoke(user string, privileges PrivilegeList) {
+	userPriv, ok := p.findUser(user)
+	if !ok {
+		// Removing privileges from a user without privileges is a no-op.
+		return
+	}
+	for _, p := range privileges {
+		if p == Privilege_ALL {
+			// TODO(marc): even though this should not happen, error out if
+			// ALL is not the only privilege being revoked.
+			userPriv.Privileges = 0
+			break
+		}
+		if userPriv.Privileges&(1<<uint32(Privilege_ALL)) != 0 {
+			// We're revoking something other than ALL, clear ALL
+			// and set all others.
+			userPriv.Privileges = 0
+			for _, v := range privilegeValues {
+				if v != Privilege_ALL {
+					userPriv.Privileges |= (1 << uint32(v))
+				}
+			}
+		}
+		// One doesn't see "AND NOT" very often.
+		userPriv.Privileges &^= (1 << uint32(p))
+	}
+	if userPriv.Privileges == 0 {
+		p.removeUser(user)
+	}
+}
+
+// Validate is called when writing a descriptor.
+func (p *PrivilegeDescriptor) Validate() error {
+	userPriv, ok := p.findUser(security.RootUser)
+	if !ok {
+		return fmt.Errorf("%s user does not have privileges", security.RootUser)
+	}
+	if userPriv.Privileges&(1<<uint32(Privilege_ALL)) == 0 {
+		return fmt.Errorf("%s user does not have ALL privileges", security.RootUser)
+	}
+	return nil
+}
+
+// UserPrivilegeString is a pair of strings describing the
+// privileges for a given user.
+type UserPrivilegeString struct {
+	User       string
+	Privileges string
+}
+
+// Show returns the list of {username, privileges} sorted by username.
+// The privileges is a comma-separated sorted string of privilege names.
```

Comment:
replaced second line with `'privileges' is a string of comma-separated sorted privilege names.`